pkg/election/election.go

```diff
@@ -255,67 +263,60 @@ func (e *Election) campaignLoop(ctx context.Context, session *concurrency.Sessio
 
 				e.l.Info("get response from election observe", zap.String("key", string(resp.Kvs[0].Key)), zap.String("value", string(resp.Kvs[0].Value)))
 				leaderKey = string(resp.Kvs[0].Key)
-				leaderInfo, err := getCompaingerInfo(resp.Kvs[0].Value)
+				leaderInfo, err = getCompaingerInfo(resp.Kvs[0].Value)
 				if err != nil {
 					// this should never happened
 					e.l.Error("fail to get leader information", zap.String("value", string(resp.Kvs[0].Value)), zap.Error(err))
 					continue
 				}
-				leaderID = leaderInfo.ID
 
-				if oldLeaderID == leaderID {
+				if oldLeaderID == leaderInfo.ID {
 					continue
 				} else {
-					oldLeaderID = leaderID
+					oldLeaderID = leaderInfo.ID
 					break observeElection
 				}
 			}
 		}
 
-		if len(leaderID) == 0 {
+		if leaderInfo == nil || len(leaderInfo.ID) == 0 {
 			cancel2()
 			compaignWg.Wait()
 			continue
 		}
 
-		if leaderID != e.info.ID {
-			e.l.Info("current member is not the leader", zap.String("current member", e.info.ID), zap.String("leader", leaderID))
-			e.isNotLeader()
+		if leaderInfo.ID != e.info.ID {
+			e.l.Info("current member is not the leader", zap.String("current member", e.info.ID), zap.String("leader", leaderInfo.ID))
+			e.notifyLeader(ctx, leaderInfo)
 			cancel2()
 			compaignWg.Wait()
 			continue
 		}
 
-		e.toBeLeader() // become the leader now
+		e.notifyLeader(ctx, leaderInfo) // become the leader now
 		e.watchLeader(ctx, session, leaderKey)
-		e.retireLeader() // need to re-campaign
+		e.l.Info("retire from leader", zap.String("current member", e.info.ID))
+		e.notifyLeader(ctx, nil) // need to re-campaign
 
 		cancel2()
 		compaignWg.Wait()
 	}
 }
 
-func (e *Election) toBeLeader() {
-	e.isLeader.Set(true)
-	select {
-	case e.leaderCh <- IsLeader:
-	default:
-	}
-}
-
-func (e *Election) retireLeader() {
-	e.isLeader.Set(false)
-	select {
-	case e.leaderCh <- RetireFromLeader:
-	default:
+// notifyLeader notify the leader's information.
+// leader info can be nil, and this is used when retire from leader.
+func (e *Election) notifyLeader(ctx context.Context, leaderInfo *CampaignerInfo) {
+	if leaderInfo != nil && leaderInfo.ID == e.info.ID {
+		e.isLeader.Set(true)
+	} else {
+		e.isLeader.Set(false)
 	}
-}
 
-func (e *Election) isNotLeader() {
-	e.isLeader.Set(false)
 	select {
-	case e.leaderCh <- IsNotLeader:
-	default:
+	case e.leaderCh <- leaderInfo:
+	case <-time.After(NotifyBlockTime):
```

Comment:
Why can't we use default here? Is to avoid lost notify when restarting the `Election` instance?

Reply:
if use default, the master may lose some notify if the master is busy with old notify, for example, start the `scheduer` when it becomes leader, it will cost some time.